### ISSUE TITLE
Resizing the pinned icon editor is super slow + Pin behavior suggestion

### ISSUE THREAD

**Issue by pushqrdx** [jmanc3/winbar#9]
The pinned icon editor resize is very slow, like window decoration resizes and i can see the UI slowly repaint (but smoothly).
___

I also have a suggestion about the pinning behavior. I think when i click pin, winbar should pin with the defaults it has _(the ones displayed in the editor)_; Then if i want to edit pinned app later, like in case of wrong options, i can right click > edit pinned app to get the editor

**Comment by jmanc3** [comment]
I haven't been able to reproduce the slow resizing with lxqt and compiz (which is what I think you're running). If that's not what you are running, let me know what you are. If it, then try some other window manager (like openbox, gnome, kde, or whatever) to see if it's something specific to lxqt or compiz.

As for the pin suggestion, in the updated winbar.cfg inside the winbar.zip there is a new variable "open_pinned_icon_editor" that has a comment above it explaining what the different options are for what happens when you try to pin an icon. By default it's set to "WHEN_ANY_FIELD_EMPTY" but you can set it to "NEVER" if you'd prefer that behavior.

**Comment by pushqrdx** [comment]
![Peek 2021-05-27 02-12](https://user-images.githubusercontent.com/25884937/119746596-18c48600-be91-11eb-9349-21e74ece4ffb.gif)

@jmanc3 here's a screen cap, and yes i am using compiz 0.9. Notice that when there was no icon resizing is perfectly smooth, but with the second edit (emacs) you can see how slow it is

**Comment by jmanc3** [comment]
I was able to replicate and profile. The problem is that the repaint is just slow. Three things need fixing.

1. We repaint too often. Right now we repaint on any and every event sent to a window and that means that (in compiz) we layout once and then repaint the same thing three times on each resize because we are sent a configure_event, a exposure_event, and then a property_notify event. (other wm's don't spam us this hard I will say). The fix is to only repaint on events that will induce a visible change.
2. Improve cairo speed. I'm fairly certain I'm not falling victim to the issues brought up in this comment https://news.ycombinator.com/item?id=16539446 yet for some reason cairo (the thing used for drawing) simply slows to a crawl when trying to paint moderately sized images. For instance in your gif the icon is 512x512, but from the profiling I did, just drawing that one image takes 8ms which is absolutely insane given that means it takes half the frame budget at 60 fps (16ms). I must be doing something wrong. There's no way it should be that slow.
3. Icons can still be the wrong size. I actually can't believe this is an issue. There's only two places in pinned_icon_editor.cpp where we load an image and in both we specify that the max size of the surface is 64x64.  So I have no idea how it is able to draw a surface larger than that size. I pushed a new commit. See if the pinned icon editor can still show images THAT large.
